<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content=Aquascope><meta name=author content=SciComp><link rel=icon href=../assets/images/favicon.png><meta name=generator content="mkdocs-1.6.0, mkdocs-material-9.5.20"><title>nf-core/aquascope: Usage - Aquascope</title><link rel=stylesheet href=../assets/stylesheets/main.66ac8b77.min.css><link rel=stylesheet href=../assets/stylesheets/palette.06af60db.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#nf-coreaquascope-usage class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <div data-md-color-scheme=default data-md-component=outdated hidden> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=.. title=Aquascope class="md-header__button md-logo" aria-label=Aquascope data-md-component=logo> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Aquascope </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> nf-core/aquascope: Usage </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=red data-md-color-accent=red aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg> </label> </form> <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/CDCgov/Aquascope title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> CDCgov/Aquascope </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=.. class=md-tabs__link> Background </a> </li> <li class=md-tabs__item> <a href=../user-guide/getting-started/ class=md-tabs__link> Usage </a> </li> <li class=md-tabs__item> <a href=../user-guide/troubleshooting/ class=md-tabs__link> FAQ </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=.. title=Aquascope class="md-nav__button md-logo" aria-label=Aquascope data-md-component=logo> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg> </a> Aquascope </label> <div class=md-nav__source> <a href=https://github.com/CDCgov/Aquascope title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> CDCgov/Aquascope </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=.. class=md-nav__link> <span class=md-ellipsis> Background </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex=0> <span class=md-ellipsis> Usage </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=false> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Usage </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../user-guide/getting-started/ class=md-nav__link> <span class=md-ellipsis> 1. Getting Started </span> </a> </li> <li class=md-nav__item> <a href=../user-guide/preparing-files/ class=md-nav__link> <span class=md-ellipsis> 2. Preparing Files </span> </a> </li> <li class=md-nav__item> <a href=../user-guide/run/ class=md-nav__link> <span class=md-ellipsis> 3. Running the Pipeline </span> </a> </li> <li class=md-nav__item> <a href=../user-guide/output/ class=md-nav__link> <span class=md-ellipsis> 4. Expected Output </span> </a> </li> <li class=md-nav__item> <a href=../user-guide/test-info/ class=md-nav__link> <span class=md-ellipsis> 5. Running Test Data </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex=0> <span class=md-ellipsis> FAQ </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> FAQ </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../user-guide/troubleshooting/ class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> </li> <li class=md-nav__item> <a href=../user-guide/contributions/ class=md-nav__link> <span class=md-ellipsis> Contributions </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=nf-coreaquascope-usage>nf-core/aquascope: Usage<a class=headerlink href=#nf-coreaquascope-usage title="Permanent link">&para;</a></h1> <h2 id=please-read-this-documentation-on-the-nf-core-website-httpsnf-coreaquascopeusage><img alt=âš  class=twemoji src=https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.0.3/assets/svg/26a0.svg title=:warning:> Please read this documentation on the nf-core website: <a href=https://nf-co.re/aquascope/usage>https://nf-co.re/aquascope/usage</a><a class=headerlink href=#please-read-this-documentation-on-the-nf-core-website-httpsnf-coreaquascopeusage title="Permanent link">&para;</a></h2> <blockquote> <p><em>Documentation of pipeline parameters is generated automatically from the pipeline schema and can no longer be found in markdown files.</em></p> </blockquote> <h2 id=introduction>Introduction<a class=headerlink href=#introduction title="Permanent link">&para;</a></h2> <!-- TODO nf-core: Add documentation about anything specific to running your pipeline. For general topics, please point to (and add to) the main nf-core website. --> <h2 id=samplesheet-input>Samplesheet input<a class=headerlink href=#samplesheet-input title="Permanent link">&para;</a></h2> <p>You will need to create a samplesheet with information about the samples you would like to analyse before running the pipeline. Use this parameter to specify its location. It has to be a comma-separated file with 3 columns, and a header row as shown in the examples below.</p> <div class=highlight><pre><span></span><code><span class=go>--input &#39;[path to samplesheet file]&#39;</span>
</code></pre></div> <h3 id=multiple-runs-of-the-same-sample>Multiple runs of the same sample<a class=headerlink href=#multiple-runs-of-the-same-sample title="Permanent link">&para;</a></h3> <p>The <code>sample</code> identifiers have to be the same when you have re-sequenced the same sample more than once e.g. to increase sequencing depth. The pipeline will concatenate the raw reads before performing any downstream analysis. Below is an example for the same sample sequenced across 3 lanes:</p> <div class=highlight><pre><span></span><code><span class=go>sample,fastq_1,fastq_2</span>
<span class=go>CONTROL_REP1,AEG588A1_S1_L002_R1_001.fastq.gz,AEG588A1_S1_L002_R2_001.fastq.gz</span>
<span class=go>CONTROL_REP1,AEG588A1_S1_L003_R1_001.fastq.gz,AEG588A1_S1_L003_R2_001.fastq.gz</span>
<span class=go>CONTROL_REP1,AEG588A1_S1_L004_R1_001.fastq.gz,AEG588A1_S1_L004_R2_001.fastq.gz</span>
</code></pre></div> <h3 id=full-samplesheet>Full samplesheet<a class=headerlink href=#full-samplesheet title="Permanent link">&para;</a></h3> <p>The pipeline will auto-detect whether a sample is single- or paired-end using the information provided in the samplesheet. The samplesheet can have as many columns as you desire, however, there is a strict requirement for the first 3 columns to match those defined in the table below.</p> <p>A final samplesheet file consisting of both single- and paired-end data may look something like the one below. This is for 6 samples, where <code>TREATMENT_REP3</code> has been sequenced twice.</p> <div class=highlight><pre><span></span><code><span class=go>sample,fastq_1,fastq_2</span>
<span class=go>CONTROL_REP1,AEG588A1_S1_L002_R1_001.fastq.gz,AEG588A1_S1_L002_R2_001.fastq.gz</span>
<span class=go>CONTROL_REP2,AEG588A2_S2_L002_R1_001.fastq.gz,AEG588A2_S2_L002_R2_001.fastq.gz</span>
<span class=go>CONTROL_REP3,AEG588A3_S3_L002_R1_001.fastq.gz,AEG588A3_S3_L002_R2_001.fastq.gz</span>
<span class=go>TREATMENT_REP1,AEG588A4_S4_L003_R1_001.fastq.gz,</span>
<span class=go>TREATMENT_REP2,AEG588A5_S5_L003_R1_001.fastq.gz,</span>
<span class=go>TREATMENT_REP3,AEG588A6_S6_L003_R1_001.fastq.gz,</span>
<span class=go>TREATMENT_REP3,AEG588A6_S6_L004_R1_001.fastq.gz,</span>
</code></pre></div> <table> <thead> <tr> <th>Column</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>sample</code></td> <td>Custom sample name. This entry will be identical for multiple sequencing libraries/runs from the same sample. Spaces in sample names are automatically converted to underscores (<code>_</code>).</td> </tr> <tr> <td><code>fastq_1</code></td> <td>Full path to FastQ file for Illumina short reads 1. File has to be gzipped and have the extension ".fastq.gz" or ".fq.gz".</td> </tr> <tr> <td><code>fastq_2</code></td> <td>Full path to FastQ file for Illumina short reads 2. File has to be gzipped and have the extension ".fastq.gz" or ".fq.gz".</td> </tr> </tbody> </table> <p>An <a href=../assets/samplesheet.csv>example samplesheet</a> has been provided with the pipeline.</p> <h2 id=running-the-pipeline>Running the pipeline<a class=headerlink href=#running-the-pipeline title="Permanent link">&para;</a></h2> <p>The typical command for running the pipeline is as follows:</p> <div class=highlight><pre><span></span><code><span class=go>nextflow run nf-core/aquascope --input samplesheet.csv --genome GRCh37 -profile docker</span>
</code></pre></div> <p>This will launch the pipeline with the <code>docker</code> configuration profile. See below for more information about profiles.</p> <p>Note that the pipeline will create the following files in your working directory:</p> <div class=highlight><pre><span></span><code><span class=go>work            # Directory containing the nextflow working files</span>
<span class=go>results         # Finished results (configurable, see below)</span>
<span class=go>.nextflow_log   # Log file from Nextflow</span>
<span class=gp># </span>Other<span class=w> </span>nextflow<span class=w> </span>hidden<span class=w> </span>files,<span class=w> </span>eg.<span class=w> </span><span class=nb>history</span><span class=w> </span>of<span class=w> </span>pipeline<span class=w> </span>runs<span class=w> </span>and<span class=w> </span>old<span class=w> </span>logs.
</code></pre></div> <h3 id=updating-the-pipeline>Updating the pipeline<a class=headerlink href=#updating-the-pipeline title="Permanent link">&para;</a></h3> <p>When you run the above command, Nextflow automatically pulls the pipeline code from GitHub and stores it as a cached version. When running the pipeline after this, it will always use the cached version if available - even if the pipeline has been updated since. To make sure that you're running the latest version of the pipeline, make sure that you regularly update the cached version of the pipeline:</p> <div class=highlight><pre><span></span><code><span class=go>nextflow pull nf-core/aquascope</span>
</code></pre></div> <h3 id=reproducibility>Reproducibility<a class=headerlink href=#reproducibility title="Permanent link">&para;</a></h3> <p>It is a good idea to specify a pipeline version when running the pipeline on your data. This ensures that a specific version of the pipeline code and software are used when you run your pipeline. If you keep using the same tag, you'll be running the same version of the pipeline, even if there have been changes to the code since.</p> <p>First, go to the <a href=https://github.com/nf-core/aquascope/releases>nf-core/aquascope releases page</a> and find the latest version number - numeric only (eg. <code>1.3.1</code>). Then specify this when running the pipeline with <code>-r</code> (one hyphen) - eg. <code>-r 1.3.1</code>.</p> <p>This version number will be logged in reports when you run the pipeline, so that you'll know what you used when you look back in the future.</p> <h2 id=core-nextflow-arguments>Core Nextflow arguments<a class=headerlink href=#core-nextflow-arguments title="Permanent link">&para;</a></h2> <blockquote> <p><strong>NB:</strong> These options are part of Nextflow and use a <em>single</em> hyphen (pipeline parameters use a double-hyphen).</p> </blockquote> <h3 id=-profile><code>-profile</code><a class=headerlink href=#-profile title="Permanent link">&para;</a></h3> <p>Use this parameter to choose a configuration profile. Profiles can give configuration presets for different compute environments.</p> <p>Several generic profiles are bundled with the pipeline which instruct the pipeline to use software packaged using different methods (Docker, Singularity, Podman, Shifter, Charliecloud, Conda) - see below. When using Biocontainers, most of these software packaging methods pull Docker containers from quay.io e.g <a href=https://quay.io/repository/biocontainers/fastqc>FastQC</a> except for Singularity which directly downloads Singularity images via https hosted by the <a href=https://depot.galaxyproject.org/singularity/ >Galaxy project</a> and Conda which downloads and installs software locally from <a href=https://bioconda.github.io/ >Bioconda</a>.</p> <blockquote> <p>We highly recommend the use of Docker or Singularity containers for full pipeline reproducibility, however when this is not possible, Conda is also supported.</p> </blockquote> <p>The pipeline also dynamically loads configurations from <a href=https://github.com/nf-core/configs>https://github.com/nf-core/configs</a> when it runs, making multiple config profiles for various institutional clusters available at run time. For more information and to see if your system is available in these configs please see the <a href=https://github.com/nf-core/configs#documentation>nf-core/configs documentation</a>.</p> <p>Note that multiple profiles can be loaded, for example: <code>-profile test,docker</code> - the order of arguments is important! They are loaded in sequence, so later profiles can overwrite earlier profiles.</p> <p>If <code>-profile</code> is not specified, the pipeline will run locally and expect all software to be installed and available on the <code>PATH</code>. This is <em>not</em> recommended.</p> <ul> <li><code>docker</code><ul> <li>A generic configuration profile to be used with <a href=https://docker.com/ >Docker</a></li> </ul> </li> <li><code>singularity</code><ul> <li>A generic configuration profile to be used with <a href=https://sylabs.io/docs/ >Singularity</a></li> </ul> </li> <li><code>podman</code><ul> <li>A generic configuration profile to be used with <a href=https://podman.io/ >Podman</a></li> </ul> </li> <li><code>shifter</code><ul> <li>A generic configuration profile to be used with <a href=https://nersc.gitlab.io/development/shifter/how-to-use/ >Shifter</a></li> </ul> </li> <li><code>charliecloud</code><ul> <li>A generic configuration profile to be used with <a href=https://hpc.github.io/charliecloud/ >Charliecloud</a></li> </ul> </li> <li><code>conda</code><ul> <li>A generic configuration profile to be used with <a href=https://conda.io/docs/ >Conda</a>. Please only use Conda as a last resort i.e. when it's not possible to run the pipeline with Docker, Singularity, Podman, Shifter or Charliecloud.</li> </ul> </li> <li><code>test</code><ul> <li>A profile with a complete configuration for automated testing</li> <li>Includes links to test data so needs no other parameters</li> </ul> </li> </ul> <h3 id=-resume><code>-resume</code><a class=headerlink href=#-resume title="Permanent link">&para;</a></h3> <p>Specify this when restarting a pipeline. Nextflow will used cached results from any pipeline steps where the inputs are the same, continuing from where it got to previously.</p> <p>You can also supply a run name to resume a specific run: <code>-resume [run-name]</code>. Use the <code>nextflow log</code> command to show previous run names.</p> <h3 id=-c><code>-c</code><a class=headerlink href=#-c title="Permanent link">&para;</a></h3> <p>Specify the path to a specific config file (this is a core Nextflow command). See the <a href=https://nf-co.re/usage/configuration>nf-core website documentation</a> for more information.</p> <h2 id=custom-configuration>Custom configuration<a class=headerlink href=#custom-configuration title="Permanent link">&para;</a></h2> <h3 id=resource-requests>Resource requests<a class=headerlink href=#resource-requests title="Permanent link">&para;</a></h3> <p>Whilst the default requirements set within the pipeline will hopefully work for most people and with most input data, you may find that you want to customise the compute resources that the pipeline requests. Each step in the pipeline has a default set of requirements for number of CPUs, memory and time. For most of the steps in the pipeline, if the job exits with any of the error codes specified <a href=https://github.com/nf-core/rnaseq/blob/4c27ef5610c87db00c3c5a3eed10b1d161abf575/conf/base.config#L18>here</a> it will automatically be resubmitted with higher requests (2 x original, then 3 x original). If it still fails after the third attempt then the pipeline execution is stopped.</p> <p>For example, if the nf-core/rnaseq pipeline is failing after multiple re-submissions of the <code>STAR_ALIGN</code> process due to an exit code of <code>137</code> this would indicate that there is an out of memory issue:</p> <div class=highlight><pre><span></span><code><span class=go>[62/149eb0] NOTE: Process `RNASEQ:ALIGN_STAR:STAR_ALIGN (WT_REP1)` terminated with an error exit status (137) -- Execution is retried (1)</span>
<span class=go>Error executing process &gt; &#39;RNASEQ:ALIGN_STAR:STAR_ALIGN (WT_REP1)&#39;</span>

<span class=go>Caused by:</span>
<span class=go>    Process `RNASEQ:ALIGN_STAR:STAR_ALIGN (WT_REP1)` terminated with an error exit status (137)</span>

<span class=go>Command executed:</span>
<span class=go>    STAR \</span>
<span class=go>        --genomeDir star \</span>
<span class=go>        --readFilesIn WT_REP1_trimmed.fq.gz  \</span>
<span class=go>        --runThreadN 2 \</span>
<span class=go>        --outFileNamePrefix WT_REP1. \</span>
<span class=go>        &lt;TRUNCATED&gt;</span>

<span class=go>Command exit status:</span>
<span class=go>    137</span>

<span class=go>Command output:</span>
<span class=go>    (empty)</span>

<span class=go>Command error:</span>
<span class=go>    .command.sh: line 9:  30 Killed    STAR --genomeDir star --readFilesIn WT_REP1_trimmed.fq.gz --runThreadN 2 --outFileNamePrefix WT_REP1. &lt;TRUNCATED&gt;</span>
<span class=go>Work dir:</span>
<span class=go>    /home/pipelinetest/work/9d/172ca5881234073e8d76f2a19c88fb</span>

<span class=go>Tip: you can replicate the issue by changing to the process work dir and entering the command `bash .command.run`</span>
</code></pre></div> <p>To bypass this error you would need to find exactly which resources are set by the <code>STAR_ALIGN</code> process. The quickest way is to search for <code>process STAR_ALIGN</code> in the <a href="https://github.com/nf-core/rnaseq/search?q=process+STAR_ALIGN">nf-core/rnaseq Github repo</a>. We have standardised the structure of Nextflow DSL2 pipelines such that all module files will be present in the <code>modules/</code> directory and so based on the search results the file we want is <code>modules/nf-core/software/star/align/main.nf</code>. If you click on the link to that file you will notice that there is a <code>label</code> directive at the top of the module that is set to <a href=https://github.com/nf-core/rnaseq/blob/4c27ef5610c87db00c3c5a3eed10b1d161abf575/modules/nf-core/software/star/align/main.nf#L9><code>label process_high</code></a>. The <a href=https://www.nextflow.io/docs/latest/process.html#label>Nextflow <code>label</code></a> directive allows us to organise workflow processes in separate groups which can be referenced in a configuration file to select and configure subset of processes having similar computing requirements. The default values for the <code>process_high</code> label are set in the pipeline's <a href=https://github.com/nf-core/rnaseq/blob/4c27ef5610c87db00c3c5a3eed10b1d161abf575/conf/base.config#L33-L37><code>base.config</code></a> which in this case is defined as 72GB. Providing you haven't set any other standard nf-core parameters to <strong>cap</strong> the <a href=https://nf-co.re/usage/configuration#max-resources>maximum resources</a> used by the pipeline then we can try and bypass the <code>STAR_ALIGN</code> process failure by creating a custom config file that sets at least 72GB of memory, in this case increased to 100GB. The custom config below can then be provided to the pipeline via the <a href=#-c><code>-c</code></a> parameter as highlighted in previous sections.</p> <div class=highlight><pre><span></span><code>process {
    withName: STAR_ALIGN {
        memory = 100.GB
    }
}
</code></pre></div> <blockquote> <p><strong>NB:</strong> We specify just the process name i.e. <code>STAR_ALIGN</code> in the config file and not the full task name string that is printed to screen in the error message or on the terminal whilst the pipeline is running i.e. <code>RNASEQ:ALIGN_STAR:STAR_ALIGN</code>. You may get a warning suggesting that the process selector isn't recognised but you can ignore that if the process name has been specified correctly. This is something that needs to be fixed upstream in core Nextflow.</p> </blockquote> <h3 id=tool-specific-options>Tool-specific options<a class=headerlink href=#tool-specific-options title="Permanent link">&para;</a></h3> <p>For the ultimate flexibility, we have implemented and are using Nextflow DSL2 modules in a way where it is possible for both developers and users to change tool-specific command-line arguments (e.g. providing an additional command-line argument to the <code>STAR_ALIGN</code> process) as well as publishing options (e.g. saving files produced by the <code>STAR_ALIGN</code> process that aren't saved by default by the pipeline). In the majority of instances, as a user you won't have to change the default options set by the pipeline developer(s), however, there may be edge cases where creating a simple custom config file can improve the behaviour of the pipeline if for example it is failing due to a weird error that requires setting a tool-specific parameter to deal with smaller / larger genomes.</p> <p>The command-line arguments passed to STAR in the <code>STAR_ALIGN</code> module are a combination of:</p> <ul> <li> <p>Mandatory arguments or those that need to be evaluated within the scope of the module, as supplied in the <a href=https://github.com/nf-core/rnaseq/blob/4c27ef5610c87db00c3c5a3eed10b1d161abf575/modules/nf-core/software/star/align/main.nf#L49-L55><code>script</code></a> section of the module file.</p> </li> <li> <p>An <a href=https://github.com/nf-core/rnaseq/blob/4c27ef5610c87db00c3c5a3eed10b1d161abf575/modules/nf-core/software/star/align/main.nf#L56><code>options.args</code></a> string of non-mandatory parameters that is set to be empty by default in the module but can be overwritten when including the module in the sub-workflow / workflow context via the <code>addParams</code> Nextflow option.</p> </li> </ul> <p>The nf-core/rnaseq pipeline has a sub-workflow (see <a href=https://github.com/nf-core/modules#terminology>terminology</a>) specifically to align reads with STAR and to sort, index and generate some basic stats on the resulting BAM files using SAMtools. At the top of this file we import the <code>STAR_ALIGN</code> module via the Nextflow <a href=https://github.com/nf-core/rnaseq/blob/4c27ef5610c87db00c3c5a3eed10b1d161abf575/subworkflows/nf-core/align_star.nf#L10><code>include</code></a> keyword and by default the options passed to the module via the <code>addParams</code> option are set as an empty Groovy map <a href=https://github.com/nf-core/rnaseq/blob/4c27ef5610c87db00c3c5a3eed10b1d161abf575/subworkflows/nf-core/align_star.nf#L5>here</a>; this in turn means <code>options.args</code> will be set to empty by default in the module file too. This is an intentional design choice and allows us to implement well-written sub-workflows composed of a chain of tools that by default run with the bare minimum parameter set for any given tool in order to make it much easier to share across pipelines and to provide the flexibility for users and developers to customise any non-mandatory arguments.</p> <p>When including the sub-workflow above in the main pipeline workflow we use the same <code>include</code> statement, however, we now have the ability to overwrite options for each of the tools in the sub-workflow including the <a href=https://github.com/nf-core/rnaseq/blob/4c27ef5610c87db00c3c5a3eed10b1d161abf575/workflows/rnaseq.nf#L225><code>align_options</code></a> variable that will be used specifically to overwrite the optional arguments passed to the <code>STAR_ALIGN</code> module. In this case, the options to be provided to <code>STAR_ALIGN</code> have been assigned sensible defaults by the developer(s) in the pipeline's <a href=https://github.com/nf-core/rnaseq/blob/4c27ef5610c87db00c3c5a3eed10b1d161abf575/conf/modules.config#L70-L74><code>modules.config</code></a> and can be accessed and customised in the <a href=https://github.com/nf-core/rnaseq/blob/4c27ef5610c87db00c3c5a3eed10b1d161abf575/workflows/rnaseq.nf#L201-L204>workflow context</a> too before eventually passing them to the sub-workflow as a Groovy map called <code>star_align_options</code>. These options will then be propagated from <code>workflow -&gt; sub-workflow -&gt; module</code>.</p> <p>As mentioned at the beginning of this section it may also be necessary for users to overwrite the options passed to modules to be able to customise specific aspects of the way in which a particular tool is executed by the pipeline. Given that all of the default module options are stored in the pipeline's <code>modules.config</code> as a <a href=https://github.com/nf-core/rnaseq/blob/4c27ef5610c87db00c3c5a3eed10b1d161abf575/conf/modules.config#L24-L25><code>params</code> variable</a> it is also possible to overwrite any of these options via a custom config file.</p> <p>Say for example we want to append an additional, non-mandatory parameter (i.e. <code>--outFilterMismatchNmax 16</code>) to the arguments passed to the <code>STAR_ALIGN</code> module. Firstly, we need to copy across the default <code>args</code> specified in the <a href=https://github.com/nf-core/rnaseq/blob/4c27ef5610c87db00c3c5a3eed10b1d161abf575/conf/modules.config#L71><code>modules.config</code></a> and create a custom config file that is a composite of the default <code>args</code> as well as the additional options you would like to provide. This is very important because Nextflow will overwrite the default value of <code>args</code> that you provide via the custom config.</p> <p>As you will see in the example below, we have:</p> <ul> <li>appended <code>--outFilterMismatchNmax 16</code> to the default <code>args</code> used by the module.</li> <li>changed the default <code>publish_dir</code> value to where the files will eventually be published in the main results directory.</li> <li>appended <code>'bam':''</code> to the default value of <code>publish_files</code> so that the BAM files generated by the process will also be saved in the top-level results directory for the module. Note: <code>'out':'log'</code> means any file/directory ending in <code>out</code> will now be saved in a separate directory called <code>my_star_directory/log/</code>.</li> </ul> <div class=highlight><pre><span></span><code>params {
    modules {
        &#39;star_align&#39; {
            args          = &quot;--quantMode TranscriptomeSAM --twopassMode Basic --outSAMtype BAM Unsorted --readFilesCommand zcat --runRNGseed 0 --outFilterMultimapNmax 20 --alignSJDBoverhangMin 1 --outSAMattributes NH HI AS NM MD --quantTranscriptomeBan Singleend --outFilterMismatchNmax 16&quot;
            publish_dir   = &quot;my_star_directory&quot;
            publish_files = [&#39;out&#39;:&#39;log&#39;, &#39;tab&#39;:&#39;log&#39;, &#39;bam&#39;:&#39;&#39;]
        }
    }
}
</code></pre></div> <h3 id=updating-containers>Updating containers<a class=headerlink href=#updating-containers title="Permanent link">&para;</a></h3> <p>The <a href=https://www.nextflow.io/docs/latest/dsl2.html>Nextflow DSL2</a> implementation of this pipeline uses one container per process which makes it much easier to maintain and update software dependencies. If for some reason you need to use a different version of a particular tool with the pipeline then you just need to identify the <code>process</code> name and override the Nextflow <code>container</code> definition for that process using the <code>withName</code> declaration. For example, in the <a href=https://nf-co.re/viralrecon>nf-core/viralrecon</a> pipeline a tool called <a href=https://github.com/cov-lineages/pangolin>Pangolin</a> has been used during the COVID-19 pandemic to assign lineages to SARS-CoV-2 genome sequenced samples. Given that the lineage assignments change quite frequently it doesn't make sense to re-release the nf-core/viralrecon everytime a new version of Pangolin has been released. However, you can override the default container used by the pipeline by creating a custom config file and passing it as a command-line argument via <code>-c custom.config</code>.</p> <ol> <li>Check the default version used by the pipeline in the module file for <a href=https://github.com/nf-core/viralrecon/blob/a85d5969f9025409e3618d6c280ef15ce417df65/modules/nf-core/software/pangolin/main.nf#L14-L19>Pangolin</a></li> <li>Find the latest version of the Biocontainer available on <a href="https://quay.io/repository/biocontainers/pangolin?tag=latest&tab=tags">Quay.io</a></li> <li> <p>Create the custom config accordingly:</p> <ul> <li> <p>For Docker:</p> <div class=highlight><pre><span></span><code>process {
    withName: PANGOLIN {
        container = &#39;quay.io/biocontainers/pangolin:3.0.5--pyhdfd78af_0&#39;
    }
}
</code></pre></div> </li> <li> <p>For Singularity:</p> <div class=highlight><pre><span></span><code>process {
    withName: PANGOLIN {
        container = &#39;https://depot.galaxyproject.org/singularity/pangolin:3.0.5--pyhdfd78af_0&#39;
    }
}
</code></pre></div> </li> <li> <p>For Conda:</p> <div class=highlight><pre><span></span><code>process {
    withName: PANGOLIN {
        conda = &#39;bioconda::pangolin=3.0.5&#39;
    }
}
</code></pre></div> </li> </ul> </li> </ol> <blockquote> <p><strong>NB:</strong> If you wish to periodically update individual tool-specific results (e.g. Pangolin) generated by the pipeline then you must ensure to keep the <code>work/</code> directory otherwise the <code>-resume</code> ability of the pipeline will be compromised and it will restart from scratch.</p> </blockquote> <h3 id=nf-coreconfigs>nf-core/configs<a class=headerlink href=#nf-coreconfigs title="Permanent link">&para;</a></h3> <p>In most cases, you will only need to create a custom config as a one-off but if you and others within your organisation are likely to be running nf-core pipelines regularly and need to use the same settings regularly it may be a good idea to request that your custom config file is uploaded to the <code>nf-core/configs</code> git repository. Before you do this please can you test that the config file works with your pipeline of choice using the <code>-c</code> parameter. You can then create a pull request to the <code>nf-core/configs</code> repository with the addition of your config file, associated documentation file (see examples in <a href=https://github.com/nf-core/configs/tree/master/docs><code>nf-core/configs/docs</code></a>), and amending <a href=https://github.com/nf-core/configs/blob/master/nfcore_custom.config><code>nfcore_custom.config</code></a> to include your custom profile.</p> <p>See the main <a href=https://www.nextflow.io/docs/latest/config.html>Nextflow documentation</a> for more information about creating your own configuration files.</p> <p>If you have any questions or issues please send us a message on <a href=https://nf-co.re/join/slack>Slack</a> on the <a href=https://nfcore.slack.com/channels/configs><code>#configs</code> channel</a>.</p> <h2 id=running-in-the-background>Running in the background<a class=headerlink href=#running-in-the-background title="Permanent link">&para;</a></h2> <p>Nextflow handles job submissions and supervises the running jobs. The Nextflow process must run until the pipeline is finished.</p> <p>The Nextflow <code>-bg</code> flag launches Nextflow in the background, detached from your terminal so that the workflow does not stop if you log out of your session. The logs are saved to a file.</p> <p>Alternatively, you can use <code>screen</code> / <code>tmux</code> or similar tool to create a detached session which you can log back into at a later time. Some HPC setups also allow you to run nextflow within a cluster job submitted your job scheduler (from where it submits more jobs).</p> <h2 id=nextflow-memory-requirements>Nextflow memory requirements<a class=headerlink href=#nextflow-memory-requirements title="Permanent link">&para;</a></h2> <p>In some cases, the Nextflow Java virtual machines can start to request a large amount of memory. We recommend adding the following line to your environment to limit this (typically in <code>~/.bashrc</code> or <code>~./bash_profile</code>):</p> <div class=highlight><pre><span></span><code><span class=go>NXF_OPTS=&#39;-Xms1g -Xmx4g&#39;</span>
</code></pre></div> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1-2.1-2M12.5 7v5.2l4 2.4-1 1L11 13V7h1.5M11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2v1.8Z"/></svg> </span> 2023-06-27 </span> </aside> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2024 CDC SciComp </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://www.cdc.gov/amd/index.html target=_blank rel=noopener title=www.cdc.gov class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M144 0a80 80 0 1 1 0 160 80 80 0 1 1 0-160zm368 0a80 80 0 1 1 0 160 80 80 0 1 1 0-160zM0 298.7C0 239.8 47.8 192 106.7 192h42.7c15.9 0 31 3.5 44.6 9.7-1.3 7.2-1.9 14.7-1.9 22.3 0 38.2 16.8 72.5 43.3 96H21.3C9.6 320 0 310.4 0 298.7zM405.3 320h-.7c26.6-23.5 43.3-57.8 43.3-96 0-7.6-.7-15-1.9-22.3 13.6-6.3 28.7-9.7 44.6-9.7h42.7c58.9 0 106.7 47.8 106.7 106.7 0 11.8-9.6 21.3-21.3 21.3H405.3zM224 224a96 96 0 1 1 192 0 96 96 0 1 1-192 0zm-96 261.3c0-73.6 59.7-133.3 133.3-133.3h117.4c73.6 0 133.3 59.7 133.3 133.3 0 14.7-11.9 26.7-26.7 26.7H154.7c-14.7 0-26.7-11.9-26.7-26.7z"/></svg> </a> <a href=https://github.com/CDCgov target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "..", "features": ["navigation.tabs", "navigation.top", "navigation.indexes", "toc.integrate"], "search": "../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script> <script src=../assets/javascripts/bundle.dd8806f2.min.js></script> </body> </html>