This project is a successor to the [C-WAP pipeline](https://github.com/CFSAN-Biostatistics/C-WAP) and is intended to process SARS-CoV-2 wastewater samples to determine relative variant abundance.  

⚠️Warning⚠️ 

	The results generated by this pipeline are not CLIA certified and should not be considered diagnostic.

## Introduction
**CDCgov/aquascope** is a bioinformatics best-practice pipeline for early detection of SARS-COV variants of concern, sequenced throughshotgun metagenomic sequencing, from wastewater.

The pipeline is built using [Nextflow](https://www.nextflow.io), a workflow tool to run tasks across multiple compute infrastructures in a very portable manner. It uses Docker/Singularity containers making installation trivial and results highly reproducible. 

## Pipeline summary

1. Read QC: [`FastQC`](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
2. Trimming reads: [`Fastp`](https://github.com/OpenGene/fastp)
3. Aligning short reads: [`Minimap2`](https://github.com/lh3/minimap2)
4. Freyja Variant classification: [`Freyja`](https://github.com/andersen-lab/Freyja)
5. Present QC for raw reads: [`MultiQC`](http://multiqc.info/)